go/private/actions/link.bzl

```diff
@@ -67,23 +91,9 @@ def emit_link(
     if go.mode.link == LINKMODE_PLUGIN:
         tool_args.add_all(["-pluginpath", archive.data.importpath])
 
-    # Build the set of transitive dependencies. Currently, we tolerate multiple
-    # archives with the same importmap (though this will be an error in the
-    # future), but there is a special case which is difficult to avoid:
-    # If a go_test has internal and external archives, and the external test
-    # transitively depends on the library under test, we need to exclude the
-    # library under test and use the internal test archive instead.
-    deps = depset(transitive = [d.transitive for d in archive.direct])
-    dep_args = [
-        "{}={}={}".format(d.label, d.importmap, d.file.path)
-        for d in deps.to_list()
-        if not any([d.importmap == t.importmap for t in test_archives])
-    ]
-    dep_args.extend([
-        "{}={}={}".format(d.label, d.importmap, d.file.path)
-        for d in test_archives
-    ])
-    builder_args.add_all(dep_args, before_each = "-dep")
+    builder_args.add_all([struct(archive = archive, test_archives = test_archives)], before_each = "-dep",
+        map_each = _map_archive)
+    builder_args.add_all(test_archives, before_each = "-dep", map_each = _map_test_archive)
```

Comment:
Instead of doing the work inline here, we postpone it. It will only be done if the action needs to be run.